src/components/primitives/Text/Clamp.js

```diff
@@ -0,0 +1,149 @@
+import React, { PureComponent, Fragment } from "react";
+import { StyleSheet, Text, View } from "react-native";
+
+export default class Clamp extends PureComponent {
+  state = {
+    isTruncating: null,
+    containerWidth: null,
+    preEllipsisText: null,
+    restOfText: null,
+  };
+
+  handleSingleLineLayout = event => {
+    const singleLineWidth = event.nativeEvent.layout.width;
+    const {
+      numberOfLines,
+      children: text,
+      children: { length: totalCharCount },
+    } = this.props;
+    const { containerWidth } = this.state;
+
+    const averageCharsPerLine = getAverageCharsPerLine({
+      singleLineWidth,
+      totalCharCount,
+      containerWidth,
+    });
+
+    const { preEllipsisText, restOfText } = getText({
+      text,
+      numberOfLines,
+      averageCharsPerLine,
+    });
+
+    this.setState(state => ({
+      preEllipsisText,
+      restOfText,
+      isTruncating: false,
+    }));
+  };
+
+  handleContainerLayout = ({
+    nativeEvent: {
+      layout: { width },
+    },
+  }) => {
+    if (width === this.state.containerWidth) {
+      return;
+    }
+    this.setState(state => ({
+      containerWidth: width,
+      isTruncating: true,
+    }));
+  };
+
+  renderTextOnSingleLine = () => {
+    const { numberOfLines, style, ...rest } = this.props;
+    return (
+      <View style={styles.singleLineContainer}>
+        <Text
+          onLayout={this.handleSingleLineLayout}
+          numberOfLines={1}
+          style={[style, styles.singleLineText]}
+          {...rest}
+        >
+          {this.props.children}
+        </Text>
+      </View>
+    );
+  };
+
+  render() {
+    const { numberOfLines, ...rest } = this.props;
+    const { isTruncating, preEllipsisText, restOfText } = this.state;
+
+    return (
+      <View onLayout={this.handleContainerLayout} style={styles.container}>
+        {isTruncating ? (
+          this.renderTextOnSingleLine()
+        ) : (
+          <Fragment>
+            <PreEllipsisLines {...rest} text={preEllipsisText} />
+            <EllipsisLine {...rest} text={restOfText} />
+          </Fragment>
+        )}
+      </View>
+    );
+  }
+}
+
+const PreEllipsisLines = ({ text, ...rest }) => {
+  return <Text {...rest}>{text}</Text>;
+};
+
+const EllipsisLine = ({ text, ...rest }) => {
+  return text && text.length ? (
+    <Text numberOfLines={1} {...rest}>
+      {text}
+    </Text>
+  ) : null;
+};
+
+const getAverageCharsPerLine = ({
```

Comment:
Can you add a comment for this? I find it a little trickier than it seems at first glance. Like singleLineWidth is the screen width occupied if the full text were rendered as one line. container width is width of container to wrap in. totalCharCount / singleLineWidth is the average width of a character, multiplied by the containerWidth will get you the average character count that will fit in the width of the container.

Reply:
Yeah I wanted to add some JSDoc stuff above it -- criss cross applesauce